### PR TITLE
[FIX] Unify API

### PR DESCRIPTION
**Pull request by elephaint**
This is a large refactoring PR and open for discussion. The main goal of the PR is to unify API across different model types, and unify loss functions across different loss types.

Refactoring:
- Fuses `BaseWindows`, `BaseMultivariate` and `BaseRecurrent` into `BaseModel`, removing the need for separate classes and unifying model API across different model types. Instead, this PR introduces two model attributes, yielding four possible model options: `RECURRENT` (`True`/`False`) and `MULTIVARIATE` (`True`/`False`). We currently have a model for every combination except a recurrent multivariate model (e.g. a multivariate LSTM), however this is now relatively simple to add. In addition, this change allows to have models that can be recurrent or not, or multivariate or not on-the-fly, based on users' input. This also allows for easier modelling going forward.
- Unifies model API across all models, adding missing input variables to all model types.
- Refactors losses, a.o. removing unnecessary `domain_map` functions.
- Moves `loss.domain_map` outside of models to `BaseModel`
- Moves RevINMultivariate used by `TSMixer`, `TSMixerx` and `RMoK` to `common.modules`

Features:
- All losses compatible with all types of models (e.g. univariate/multivariate, direct/recurrent) OR appropriate protection added.
- `DistributionLoss` now supports the use of `quantiles` in `predict`, allowing for easy quantile retrieval for all DistributionLosses.
- Mixture losses (`GMM`, `PMM` and `NBMM`) now support learned weights for weighted mixture distribution outputs.
- Mixture losses now support the use of `quantiles` in `predict`, allowing for easy quantile retrieval.
- Improved stability of `ISQF` by adding softplus protection around some parameters instead of using `.abs`
- Unified API for any quantile or any confidence level during predict for both point- and distribution losses.

Bug fixes:
- `MASE` loss now works.
- Added various protections around parameter combinations that are invalid (e.g. regarding losses).
- `StudentT` increase default DoF to 3 to reduce unbound variance issues.
- All models are now tested using a test function on the AirPassengers dataset; in most models we included `eval: false` on the examples whilst not having any other tests, causing most models to effectively not being tested at all.
- IQLoss doesn't give monotonic quantiles, now it does (by quantiling the quantiles)
- When training with both a conformal method and non-conformal method, the latter is also cross-validated to compute conformity scores. This is a redundant training step, and removed in this PR.

Breaking changes:
- Rewrite of all recurrent models to get rid of the quadratic (in the sequence dimension) space complexity. As a result, it is impossible to load a recurrent model from a previous version into this version.
- Recurrent models now require an `input_size` to be given.
- `TCN` and `DRNN` are now windows models, not recurrent models.

Tests:
- Added `common._model_checks.py` that includes a model testing function. This function runs on every separate model, ensuring that every model is tested on push.

Todo:
- [x] Test models on speed/scaling as compared to current implementation across a set of datasets.
- [x] Make sure docstring of all multivariate models is updated to reflect the additional inputs

### CODE REVIEW

**Comment by AzulGarza** [comment]
this is a very cool effort @elephaint. the new features look exciting (eg losses compatibility, model unification, ...) and there are a lot of bug fixes, too!

**Comment by elephaint** [comment]
> this is a very cool effort @elephaint. the new features look exciting (eg losses compatibility, model unification, ...) and there are a lot of bug fixes, too!

Thanks! At this moment the issue is mostly that there is some performance regression; little bit of work to do there....

**Comment by elephaint** [comment]
Performance tests - first picture is baseline (existing main repo), second picture is this PR.

Conclusion
- Overall all performance tests run faster, demonstrating that overall this PR improves speed of the API.
- There is some performance regression, mainly on recurrent models. This is due to the different implementation, where recurrent models are now flexibly implemented as either direct or recurrent models (they have a flag for setting this - by default they are implemented as `direct`). I'm ok with this performance regression, as it offers massive gains in space complexity - which is the key benefit / reason to use a recurrent model.

## Model performance
Screenshot 2024-10-08 174117

## Model performance 2
Screenshot 2024-10-08 174125

## Multivariate model performance
Screenshot 2024-10-08 174132

**Comment by marcopeix** [comment]
I ran my own small experiment, and models are slightly faster, with no degradation in performance. The refactoring looks good to me, but a second opinion would be great on this!

**Comment by elephaint** [comment]
> I ran my own small experiment, and models are slightly faster, with no degradation in performance. The refactoring looks good to me, but a second opinion would be great on this!

Thanks!

File: action_files/test_models/src/evaluation.py
[the remark this answers is not in the document]
Comment (elephaint) [reply action_files/test_models/src/evaluation.py:1]: Good idea, two questions / remarks:
- Splitting the files makes the tests run faster as we effectively parallelize them in circleci
- Why favour ci over circleci? (I'm ambivalent, don't know why we would prefer one over the other)

File: action_files/test_models/src/models.py
[the remark this answers is not in the document]
Comment (elephaint) [reply action_files/test_models/src/models.py:1]: Makes sense, then we can still fire up multiple runners (for the sake of keeping test time under control it makes sense to split the tests)

File: action_files/test_models/src/models2.py [hunk 1]
```diff
@@ -2,35 +2,39 @@
...
 from neuralforecast.models.lstm import LSTM
 from neuralforecast.models.dilated_rnn import DilatedRNN
-from neuralforecast.models.deepar import DeepAR
-from neuralforecast.models.mlp import MLP
-from neuralforecast.models.nhits import NHITS
-from neuralforecast.models.nbeats import NBEATS
+# from neuralforecast.models.deepar import DeepAR
+# from neuralforecast.models.mlp import MLP
+# from neuralforecast.models.nhits import NHITS
+# from neuralforecast.models.nbeats import NBEATS
```
[the remark this answers is not in the document]
Comment (elephaint) [reply action_files/test_models/src/models2.py:21]: I'm kind of treating this file also as a testing file locally, we can delete it (it's mainly so that testing locally is faster that you don't need to type all that stuff every time)

File: nbs/common.base_model.ipynb [hunk 2]
```diff
@@ -112,27 +118,91 @@
...
+    "        # Multivarariate checks\n",
+    "        if self.MULTIVARIATE and n_series is None:\n",
+    "            raise Exception(f'{type(self).__name__} is a multivariate model. Please set n_series to the number of unique time series in your dataset.')\n",
+    "        if not self.MULTIVARIATE:\n",
+    "            if n_series is not None:\n",
+    "                warnings.warn(\n",
+    "                    f'{type(self).__name__} is a univariate model. Parameter n_series is ignored.'\n",
+    "                )\n",
+    "            n_series = 1\n",
+    "        self.n_series = n_series          \n",
```
[the remark this answers is not in the document]
Comment (elephaint) [reply nbs/common.base_model.ipynb:174]: We need `n_series=1` for the univariate models in a few other places, but I can also introduce additional helper variables to deal with that situation, but that feels a bit unnecessary perhaps....

**Comment by elephaint** [comment]
Todo:

- [ ] Check correct functioning of #1233 and add a test for that functionality in the code

File: action_files/test_models/src/evaluation.py
[the remark this answers is not in the document]
Comment (elephaint) [reply action_files/test_models/src/evaluation.py:1]: I think maybe it's better to rewrite the testing scripts in a separate PR, wdyt?

File: action_files/test_models/src/models.py
[the remark this answers is not in the document]
Comment (elephaint) [reply action_files/test_models/src/models.py:1]: I think maybe it's better to rewrite the testing scripts in a separate PR, wdyt?

**Review by cchallu — APPROVED** [review]
